### PR TITLE
Added Skip Logic: Matrix parameters cannot be empty arrays

### PR DESCRIPTION
**Pull request by EmmaMunley**
# Changes

Added a check in Pipeline validation to verify that matrix parameters are not just arrays, but that they are not empty arrays. This won't change existing functionality/capabilities, but will get rid of the panic in the reconciler. 
This addresses issues: #6071 

/kind bug

# Submitter Checklist

As the author of this PR, please check off the items in this checklist:

- [x] Has [Docs](https://github.com/tektoncd/community/blob/main/standards.md#docs) included if any changes are user facing
- [x] Has [Tests](https://github.com/tektoncd/community/blob/main/standards.md#tests) included if any functionality added or changed
- [x] Follows the [commit message standard](https://github.com/tektoncd/community/blob/main/standards.md#commits)
- [x] Meets the [Tekton contributor standards](https://github.com/tektoncd/community/blob/main/standards.md) (including
  functionality, content, code)
- [x] Has a kind label. You can add one by adding a comment on this PR that contains `/kind <type>`. Valid types are bug, cleanup, design, documentation, feature, flake, misc, question, tep
- N/A Release notes block below has been updated with any user facing changes (API changes, bug fixes, changes requiring upgrade notices or deprecation warnings)
- N/A Release notes contains the string "action required" if the change requires additional action from users switching to the new release

# Release Notes

```release-note
NONE
```

### CODE REVIEW

**Review by jerop — COMMENTED** [review]
Thank you @EmmaMunley!

- Please update the documentation:  https://github.com/tektoncd/pipeline/blob/main/docs/matrix.md
- You referenced issue https://github.com/tektoncd/pipeline/issues/6137 but it should be https://github.com/tektoncd/pipeline/issues/6071

**Comment by jerop** [comment]
cc @abayer

**Comment by pritidesai** [comment]
thanks @EmmaMunley for the PR, I have updated the PR description with the right issue number.

@abayer and @jerop, please review the comment at https://github.com/tektoncd/pipeline/issues/6071#issuecomment-1424705562

**Comment by EmmaMunley** [comment]
> thanks @EmmaMunley for the PR, I have updated the PR description with the right issue number.
> 
> @abayer and @jerop, please review the comment at [#6071 (comment)](https://github.com/tektoncd/pipeline/issues/6071#issuecomment-1424705562)

Thank you @pritidesai. I've updated the matrix documentation.

File: docs/matrix.md [hunk 1]
```diff
@@ -69,6 +69,7 @@ The `Matrix` will take `Parameters` of type `"array"` only, which will be suppli
 `PipelineTask` by substituting `Parameters` of type `"string"` in the underlying `Task`.
 The names of the `Parameters` in the `Matrix` must match the names of the `Parameters`
 in the underlying `Task` that they will be substituting.
+Note: The `Paramaters` of type `array` cannot be empty.
```
Comment (pritidesai) [inline docs/matrix.md:72]: thank you @EmmaMunley for updating the doc, I appreciate it but this is not my expected behavior is. 

As a user, I would want to:

1. With a single matrix param - `skip` the pipelineTask if matrix param is an empty array. 
2. With multiple matrix params - If all params are empty array, follow (1) else create param combination with non-empty matrix params for the underlying pipelinesTask.

There is a distinction between an empty array and uninitialized array. Uninitialized array is treated as an error but empty array could have been chosen by design. 

@jerop thoughts?

Comment (jerop) [reply docs/matrix.md:72]: @pritidesai do you have an example or use case for why an empty array should be valid? tried to think of something but couldn't come up with any because I assume that a value has to be passed to the taskrun for that param

also, if there's no non-empty array, I'd expect `failure`, instead of `skipping`, because we can't create a taskrun from that `pipelinetask`

Comment (pritidesai) [reply docs/matrix.md]: thank you @jerop!

We had a discussion in the [API WG](https://docs.google.com/document/d/1XTN44T_AzAXDRHS6relY-Au2DzoaHSu2MpT0UkH01TA/edit#heading=h.ah8u3qqpv18t) with @EmmaMunley and @afrittoli on this topic.

Use case: A pipeline is generating SBOMs using many different tools and uploading them to an object store. Here, a list of SBOMs is a matrix param. It is possible that none of the tools produced SBOM for some valid reason. As a pipeline author, I would like to continue with the rest of the pipeline instead of failing the entire run. Generating these SBOMs could be just a part of one pipeline with other tasks. As a pipeline author,

One more use case I documented in the WG notes with the list of images. As a pipeline author, test the list of images which were published based on some criteria. It is possible that there were no images published. In that case, the pipeline will not run tests but will continue running the rest of the tasks of why the images failed to meet the required criteria.

Failure is fatal and will cause the entire pipeline to fail. The pipeline could have spent hours on a list of tasks. 

>also, if there's no non-empty array, I'd expect failure, instead of skipping, because we can't create a taskrun from that pipelinetask

Please correct me if I am wrong but skipping anyways does not create a `taskRun`.

Also, this might need a little more discussion.

Now, the whole purpose of `matrix` is to generate exhaustive list of combinations for a given set of parameters.

In case of a matrix with one param and that one param being empty, `matrix` generates 0 combination. So, the natural choice is to skip the pipelineTask `my-matrix-task` because pipeline controller does not have enough information to decide whether to create one taskRun or 10 taskRuns and what should be the value of that param `param1`. Now, this param `param1` could be  associated to optional or required param of the task. We could run the underlying task once with a default value of that param. And in case of a required param, fail the pipelineRun since the required param is not initialized. But its contracting `matrix` contract of initiating taskRuns based on the number of combinations of matrix params.  

```
  - name: my-matrix-task
    matrix:
      params:
        - name: param1
          value: "[]"
```

Therefore, the best way forward could be skip creating a taskRun and therefore declare that pipelineTask as skipped. 

But for a matrix with more than one params and one of the params is an empty array:

```
  - name: my-matrix-task
    matrix:
      params:
        - name: param1
          value: "[a, b, c]"
        - name: param2
          value: "[]"
```

This `matrix` can result in three combinations with `param2` as an empty string for all three combinations or pick a default value from the task for this param. Or even when a `matrix` resulted in three combinations, skip the pipelineTask `my-matrix-task` because one of the matrix params was an empty array.

And, both matrix params set to an empty array:

```
  - name: my-matrix-task
    matrix:
      params:
        - name: param1
          value: "[]"
        - name: param2
          value: "[]"
```

This article has a few examples on `matrix` configuration based on some use cases: https://ncorti.com/blog/howto-github-actions-build-matrix

To summarize, `matrix` must not cause a pipelineRun to fail when a matrix param is set to an empty array. In case of a `matrix` without `include` from [TEP-0118](https://github.com/tektoncd/community/blob/main/teps/0118-matrix-with-explicit-combinations-of-parameters.md), skip if any one of the matrix param is an empty array.

I am putting on hold until we can come to conclusion (hope very soon)

/hold

Comment (afrittoli) [reply docs/matrix.md]: Thanks, @pritidesai - it seems to me that there are two levels to the issue, and I think the decision for both should be consistent regardless of the number of matrixed params and number of zero length arrays

1.  Are empty arrays in a matrix valid? i.e. `PipelineRun` continues running or not in case of empty arrays in matrix. Because we want to support dynamic arrays in future, I think it would be preferable *not* to fail, i.e. consider an empty array a valid alternative
2. Does the `PipelineRun` controller create `TaskRuns` in case of empty arrays in matrix? If one or more dimensions are empty the controller could still create `TaskRuns` without the specific parameter and let the `TaskRun` controller decide whether to run it (if there is a default set) or not. This could be a desirable behaviour in some cases, but I feel like it could be confusing for users, so my recommendation would be to skip if any of the matrixed arrays are empty

In case of skip, we should use apply the same logic that we apply for `when` expressions - the default skip logic would be to continue with following tasks, unless the skip logic is set to branch.

I'm not sure if we support results from a task in a matrix today, but if we do, the resulting array of results should be empty, so that the following tasks could decide how to react based on the size. 
I believe we don't support the `${#array}` syntax yet to evaluate the size of an array, I believe that would be a nice addition for these cases.

Comment (jerop) [reply docs/matrix.md]: thanks for sharing the use cases and example scenarios @pritidesai!

agree with @afrittoli about consistency - let's start by skipping the pipelinetask if any param in a matrix is empty...we can revisit the idea of fanning out using non-empty params only (and ignoring the empty params) in the future so that we can keep things simple for now, what do you think @pritidesai?

@afrittoli - we do not yet support results from a matrixed pipelinetask - [TEP-0090](https://github.com/tektoncd/community/blob/main/teps/0090-matrix.md#results-from-fanned-out-pipelinetasks )

Comment (jerop) [reply docs/matrix.md]: Discussed with @pritidesai at API WG on 3/6 and agreed to move forward with @afrittoli's suggestion that we skip the PipelineTask if there's an empty array in the Matrix -- cc @EmmaMunley

Comment (pritidesai) [reply docs/matrix.md]: thanks a bunch @jerop @afrittoli and others for all the input! Very excited to see we have come to the conclusion 🎉 

To summarize:

The initial proposal was to `fail` the `pipelineRun` when a `matrix.params` has an empty array. My initial suggestion was to `skip` in two scenarios (1) a single `matrix.params` set to empty array (2) all specified `matrix.params` are set to empty array but continue execution in third scenario (3) all specified `matrix.params` are not empty i.e. some of them are initialized to a valid array. @afrittoli suggested to keep it consistent and `skip` in all three scenarios (thank you @afrittoli for the great suggestion).

@EmmaMunley please feel free to ping me if you have any questions. I will be very happy to guide you further.

File: pkg/apis/pipeline/v1/pipelinerun_types.go [hunk 2]
```diff
@@ -470,6 +470,8 @@ const (
 	// FinallyTimedOutSkip means the task was skipped because the PipelineRun has passed its Timeouts.Finally.
 	FinallyTimedOutSkip SkippingReason = "PipelineRun Finally timeout has been reached"
 	// None means the task was not skipped
+	MatrixParamsContainEmptyArrSkip SkippingReason = "PipelineRun tasks were skipped because matrix parameters contain empty arr"
```
Comment (pritidesai) [inline pkg/apis/pipeline/v1/pipelinerun_types.go:473]: at first I thought there is a typo in `Arr` in `MatrixParamsContainEmptyArrSkip` 🙃 

NIT: Suggest to keep it simple, `EmptyArrayInMatrixParams` and set it to `"PipelineRun tasks were skipped because matrix parameters contain empty array"`

File: pkg/reconciler/pipelinerun/resources/pipelinerunresolution.go [hunk 3]
```diff
@@ -388,6 +388,8 @@ func (t *ResolvedPipelineTask) skip(facts *PipelineRunFacts) TaskSkipStatus {
 		skippingReason = v1beta1.PipelineTimedOutSkip
 	case t.skipBecausePipelineRunTasksTimeoutReached(facts):
 		skippingReason = v1beta1.TasksTimedOutSkip
+	case t.skipBecauseMatrixParamsContainEmptyArr(facts):
+		skippingReason = v1beta1.TasksTimedOutSkip
```
Comment (pritidesai) [inline pkg/reconciler/pipelinerun/resources/pipelinerunresolution.go:392]: `skippingReason` is set to wrong reason here, set to `TasksTimedOutSkip` instead of `MatrixParamsContainEmptyArrSkip`

File: pkg/apis/pipeline/v1/pipelinerun_types.go (hunk 2 above)
Comment (EmmaMunley) [reply pkg/apis/pipeline/v1/pipelinerun_types.go]: Updated naming

File: pkg/apis/pipeline/v1/pipelinerun_types.go [hunk 4]
```diff
@@ -469,6 +469,8 @@ const (
 	TasksTimedOutSkip SkippingReason = "PipelineRun Tasks timeout has been reached"
 	// FinallyTimedOutSkip means the task was skipped because the PipelineRun has passed its Timeouts.Finally.
 	FinallyTimedOutSkip SkippingReason = "PipelineRun Finally timeout has been reached"
+	// EmptyArrayInMatrixParams means the task was skipped because Matrix parameters contain empty array.
+	EmptyArrayInMatrixParams SkippingReason = "PipelineRun tasks were skipped because matrix parameters contain empty array"
```
Comment (jerop) [inline pkg/apis/pipeline/v1/pipelinerun_types.go:473]:
```suggestion
	EmptyArrayInMatrixParams SkippingReason = "Matrix Parameters have an empty array"
```

File: pkg/apis/pipeline/v1beta1/pipelinerun_types.go [hunk 5]
```diff
@@ -479,6 +479,8 @@ const (
 	TasksTimedOutSkip SkippingReason = "PipelineRun Tasks timeout has been reached"
 	// FinallyTimedOutSkip means the task was skipped because the PipelineRun has passed its Timeouts.Finally.
 	FinallyTimedOutSkip SkippingReason = "PipelineRun Finally timeout has been reached"
+	// EmptyArrayInMatrixParams means the task was skipped because Matrix parameters contain empty array.
+	EmptyArrayInMatrixParams SkippingReason = "PipelineRun tasks were skipped because matrix parameters contain empty array"
```
Comment (jerop) [inline pkg/apis/pipeline/v1beta1/pipelinerun_types.go:483]:
```suggestion
	EmptyArrayInMatrixParams SkippingReason = "Matrix Parameters have an empty array"
```

File: pkg/apis/pipeline/v1/pipelinerun_types.go (hunk 4 above)
Comment (jerop) [reply pkg/apis/pipeline/v1/pipelinerun_types.go:473]: the previous reasons reference "pipelinerun tasks / pipelinerun finally" because they are talking about the timeouts set in pipelineruns

Comment (jerop) [reply pkg/apis/pipeline/v1/pipelinerun_types.go:473]: see `MissingResultsSkip` or `WhenExpressionsSkip` for other examples

Comment (EmmaMunley) [reply pkg/apis/pipeline/v1/pipelinerun_types.go]: Ah I see. Thanks!

**Comment by pritidesai** [comment]
thanks @EmmaMunley for resolving all the comments 👍 Can I please request to add an example as well? A similar example to check if the task is skipped when a matrix param is an empty array - https://github.com/tektoncd/pipeline/blob/main/examples/v1beta1/pipelineruns/pipelinerun-with-when-expressions.yaml#L83

File: examples/v1beta1/pipelineruns/alpha/pipelinerun-with-matrix.yaml [hunk 6]
```diff
@@ -51,3 +51,8 @@ spec:
             value: chrome
         taskRef:
           name: platform-browsers
+      - name: matrix-with-skipped-tasks # matrix parameters cannot have empty array values
+        matrix:
+          params:
+            - name: platform
+              value: '[]'
```
Comment (jerop) [inline examples/v1beta1/pipelineruns/alpha/pipelinerun-with-matrix.yaml:58]: this is not a fully formed pipelinetask - needs a taskspec or taskref

added a taskspec, and implemented it such that this task should fail if it's executed (to ensure that it is always skipped)
```suggestion
      - name: matrix-params-with-empty-array-skipped
        matrix:
          params:
            - name: versions
              value: []
        taskSpec:
          params:
            - name: version
          steps:
            - name: echo
              image: ubuntu
              script: exit 1
```

also, in another PR, let's add the same examples under finally 🙏🏾

File: pkg/reconciler/pipelinerun/resources/pipelinerunresolution.go [hunk 7]
```diff
@@ -388,6 +388,8 @@ func (t *ResolvedPipelineTask) skip(facts *PipelineRunFacts) TaskSkipStatus {
 		skippingReason = v1beta1.PipelineTimedOutSkip
 	case t.skipBecausePipelineRunTasksTimeoutReached(facts):
 		skippingReason = v1beta1.TasksTimedOutSkip
+	case t.skipBecauseEmptyArrayInMatrixParams(facts):
+		skippingReason = v1beta1.EmptyArrayInMatrixParams
```
Comment (jerop) [inline pkg/reconciler/pipelinerun/resources/pipelinerunresolution.go:392]: a similar check needs to be added to [IsFinallySkipped](https://github.com/tektoncd/pipeline/blob/57577006b054c3bc45f3f668dfba890533322100/pkg/reconciler/pipelinerun/resources/pipelinerunresolution.go#L513)

**Review by jerop — COMMENTED** [review]
also please update the commit message and PR description (that we're skipping, not validating)

File: pkg/reconciler/pipelinerun/resources/pipelinerunresolution.go (hunk 7 above)
Comment (EmmaMunley) [reply pkg/reconciler/pipelinerun/resources/pipelinerunresolution.go:392]: Added to [IsFinallySkipped](https://github.com/tektoncd/pipeline/blob/57577006b054c3bc45f3f668dfba890533322100/pkg/reconciler/pipelinerun/resources/pipelinerunresolution.go#L513) as well

**Comment by jerop** [comment]
@EmmaMunley please run hack/update-codegen.sh

```Diffing /home/prow/go/src/github.com/tektoncd/pipeline against freshly generated codegen
--- /home/prow/go/src/github.com/tektoncd/pipeline/docs/pipeline-api.md	2023-03-13 18:37:50.472529682 +0000
+++ /home/prow/go/src/github.com/tektoncd/pipeline/tmpdiffroot.ShZ0o8/docs/pipeline-api.md	2023-03-13 18:28:51.567470101 +0000
@@ -3811,10 +3811,7 @@ SkippingReason
 <th>Description</th>
 </tr>
 </thead>
-<tbody><tr><td><p>&#34;Matrix Parameters have an empty array&#34;</p></td>
-<td><p>EmptyArrayInMatrixParams means the task was skipped because Matrix parameters contain empty array.</p>
-</td>
-</tr><tr><td><p>&#34;PipelineRun Finally timeout has been reached&#34;</p></td>
+<tbody><tr><td><p>&#34;PipelineRun Finally timeout has been reached&#34;</p></td>
 <td><p>FinallyTimedOutSkip means the task was skipped because the PipelineRun has passed its Timeouts.Finally.</p>
 </td>
 </tr><tr><td><p>&#34;PipelineRun was gracefully cancelled&#34;</p></td>
/home/prow/go/src/github.com/tektoncd/pipeline is out of date. Please run hack/update-codegen.sh.
```

**Comment by EmmaMunley** [comment]
> please add docs for this in matrix.md

@jerop Docs have been updated

File: docs/matrix.md [hunk 8]
```diff
@@ -70,11 +70,11 @@ For more information, see [installation customizations](install.md#customizing-b
 
 The `Matrix` will take `Parameters` of type `"array"` only, which will be supplied to the
 `PipelineTask` by substituting `Parameters` of type `"string"` in the underlying `Task`.
-The names of the `Parameters` in the `Matrix` must match the names of the `Parameters`
+A `Parameter`that contains an empty array is considered invalid. The names of the
+`Parameters` in the `Matrix` must match the names of the `Parameters`
```
Comment (jerop) [inline docs/matrix.md:74]: it is invalid implies that it'd fail, let's make it clear that it'd be skipped

```suggestion
If the `Matrix` has an empty array `Parameter`, then the `PipelineTask` will be skipped.
The names of the `Parameters` in the `Matrix` must match the names of the `Parameters`
```

File: pkg/reconciler/pipelinerun/resources/pipelinerunresolution.go [hunk 9]
```diff
@@ -504,6 +506,19 @@ func (t *ResolvedPipelineTask) skipBecausePipelineRunFinallyTimeoutReached(facts
 	return false
 }
 
+// skipBecauseEmptyArrayInMatrixParams returns true if the matrix parameters contain an empty array
+func (t *ResolvedPipelineTask) skipBecauseEmptyArrayInMatrixParams(facts *PipelineRunFacts) bool {
```
Comment (Yongxuanzhang) [inline pkg/reconciler/pipelinerun/resources/pipelinerunresolution.go:510]: Is the `facts` used in this function? Maybe we can remove it if not used.

Comment (EmmaMunley) [reply pkg/reconciler/pipelinerun/resources/pipelinerunresolution.go]: Yes, good catch. Thanks!

**Comment by pritidesai** [comment]
This PR is actively reviewed and updated. It did not make it to 0.46, moving it to 0.47.

**Comment by dibyom** [comment]
Overall LGTM! Just one question around adding another unit test

File: pkg/reconciler/taskrun/validate_taskrun.go [hunk 10]
```diff
@@ -159,8 +162,12 @@ func findMissingKeys(neededKeys, providedKeys map[string][]string) map[string][]
 // ValidateResolvedTask validates that all parameters declared in the TaskSpec are present in the taskrun
 // It also validates that all parameters have values, parameter types match the specified type and
 // object params have all the keys required
-func ValidateResolvedTask(ctx context.Context, params []v1beta1.Param, matrix *v1beta1.Matrix, rtr *resources.ResolvedTask) error {
-	if err := validateParams(ctx, rtr.TaskSpec.Params, params, matrix.GetAllParams()); err != nil {
+func ValidateResolvedTask(ctx context.Context, params v1beta1.Params, matrix *v1beta1.Matrix, rtr *resources.ResolvedTask) error {
```
[the remark this answers is not in the document]
Comment (EmmaMunley) [reply pkg/reconciler/taskrun/validate_taskrun.go:166]: @dibyom Sorry, disregard. The existing logic is correct. I was confused why we were hardcoding the type string, but it's because it is checking the type for string replacements not the param type itself. Added a comment to clarify why we are checking type string.

**Review by dibyom — APPROVED** [review]

**Review by jerop — COMMENTED** [review]
/lgtm
